properties/src/main/java/com/fasterxml/jackson/dataformat/javaprop/util/JPropPathSplitter.java

```diff
@@ -157,16 +159,64 @@ public JPropNode splitAndAdd(JPropNode parent,
 
             while ((ix = key.indexOf(_pathSeparatorChar, start)) >= start) {
                 if (ix > start) { // segment before separator
+                    if (key.charAt(ix - 1) == _escapeChar) { //potentially escaped, so process slowly
+                      return _continueWithEscapes(curr, key, start, value);
+                    }
                     String segment = key.substring(start, ix);
                     curr = _addSegment(curr, segment);
                 }
                 start = ix + 1;
-                if (start == key.length()) {
+                if (start == keyLen) {
                     break;
                 }
             }
             return _lastSegment(curr, key, start, keyLen).setValue(value);
         }
+        
+        // Working character by character to handle escapes is slower 
+        // than using indexOf, so only do it if we have an escape char 
+        // before the path separator char.
+        // Note that this resets back to the previous start, so one segment 
+        // is scanned twice.
+        private JPropNode _continueWithEscapes(JPropNode parent, String key, int start, String value) {
+            JPropNode curr = parent;
+          
+            int keylen = key.length();
+            int escCount = 0;
+            
+            StringBuilder segment = new StringBuilder();
+            
+            for (int ix = start; ix < keylen; ++ix) {
+                int cc = key.charAt(ix);
+                if (cc ==_escapeChar) {
+                    escCount++;  
+                } else if (cc == _pathSeparatorChar) {
+                    if (escCount > 0) {
+                        segment.append(key, start, ix - ((escCount + 1) >> 1));
+                        if (escCount % 2 == 0) {
+                            curr = _addSegment(curr, segment.toString());
+                            segment = new StringBuilder();
+                            start = ix + 1;
+                        } else {
+                            segment.append((char) cc);
+                            start = ix + 1;
+                            escCount = 0;
+                        }
+                    } else {
+                        segment.append(key, start, ix);
+                        curr = _addSegment(curr, segment.toString());
+                        segment = new StringBuilder();
+                        start = ix + 1;
+                    }
+                } else if (escCount > 0) {
```

Comment:
Can just use `} else {` and assign `escCount = 0` -- avoids one comparison.

Reply:
Done.